OrbitBase/ThreadPool.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitBase/ThreadPool.h"
+
+#include <list>
+#include <thread>
+
+#include "OrbitBase/Logging.h"
+#include "absl/synchronization/mutex.h"
+
+namespace {
+
+class ThreadPoolImpl : public ThreadPool {
+ public:
+  explicit ThreadPoolImpl(size_t thread_pool_min_size,
+                          size_t thread_pool_max_size, size_t queue_threshold);
+
+  void Schedule(std::unique_ptr<Action> action) override;
+  void Shutdown() override;
+  void Wait() override;
+
+ private:
+  bool ActionsAvailableOrShutdownInitiated();
+  // Blocking call - returns nullptr if shutdown initiated.
+  std::unique_ptr<Action> PopAction();
+  void CreateWorker();
+  void WorkerFunction();
+
+  absl::Mutex mutex_;
+  std::list<std::unique_ptr<Action>> scheduled_actions_;
+  std::vector<std::thread> worker_threads_;
+  size_t thread_pool_max_size_;
+  size_t queue_threshold_;
+  bool shutdown_initiated_;
+};
+
+ThreadPoolImpl::ThreadPoolImpl(size_t thread_pool_min_size,
+                               size_t thread_pool_max_size,
+                               size_t queue_threshold)
+    : thread_pool_max_size_(thread_pool_max_size),
+      queue_threshold_(queue_threshold),
+      shutdown_initiated_(false) {
+  CHECK(thread_pool_min_size > 0);
+  CHECK(thread_pool_max_size >= thread_pool_min_size);
+  CHECK(queue_threshold > 0);
+
+  absl::MutexLock lock(&mutex_);
+  for (size_t i = 0; i < thread_pool_min_size; ++i) {
+    CreateWorker();
+  }
+}
+
+void ThreadPoolImpl::CreateWorker() {
+  worker_threads_.emplace_back([this] { WorkerFunction(); });
+}
+
+void ThreadPoolImpl::Schedule(std::unique_ptr<Action> action) {
+  absl::MutexLock lock(&mutex_);
+  CHECK(!shutdown_initiated_);
+  scheduled_actions_.push_back(std::move(action));
+  if (scheduled_actions_.size() >= queue_threshold_ &&
```

Comment:
Maybe I'm missing something, but what if `queue_threshold_ == 1` and I enqueue an `Action`? Wouldn't a new thread be created even when there are already threads doing nothing?

Reply:
true - fixed.